### ISSUE TITLE
unknown type int64

### ISSUE THREAD

**Issue by ntwish** [toni-moreno/snmpcollector#388]
Release information:
Version: 0.8.0
Commit: 9e2519a
Build Date: 2018/11/5 11:11:36
OS: Ubuntu 18.04.1 LTS
After update from previous version simply snmp metric:
-conversion 1 
-ID analogValue 
-FieldName analogValue 
-BaseOID .1.3.6.1.4.1.35160.1.16.1.13 
-DataSrcType Integer32 
-Scale 0.01 
-Conversion 1 
produce message(for ex.):
Error Trying to  Scale Function from unknown type int64 value: 2180
and write unscaled value to db.
snmpwalk command result:
SNMPv2-SMI::enterprises.35160.1.16.1.13.1 = INTEGER: 2180

**Comment by toni-moreno** [comment]
Hi @ntwish I will review this issue ASAP. 

Thank you to report it

**Comment by jensenja** [comment]
Hi,

I'm bumping against this as well. I have some devices I'm polling that don't use tables and only they only return single INTEGER values via v1 SNMPGET's and I need to scale these by 0.1 in order to convert `972` to `97.2` (97.2 is the correct value). This used to work just fine before the "Conversion" feature was added to the SNMP metrics - now it's really kind of confusing for cases like this.

I'm getting the exact same error as @ntwish in my logs:

```
time="2020-03-14 16:52:40" level=error msg="Error Trying to  Scale Function from unknown type int64 value: 958"
```

I get the error if I set "Conversion" to either FLOAT or INTEGER and the SMI type that I'm using as the DATASRCTYPE is INTEGER as well.

Do we know what the root cause is? For now I'll have to roll back to the previous version of snmpcollector that I was using before the Conversion feature was implemented.

**Comment by jensenja** [comment]
I'm not sure if this is the root cause, but in `snmpmetric.go`:

```
	case "INTEGER", "Integer32":
		s.SetRawData = func(pdu gosnmp.SnmpPDU, now time.Time) {
			s.CookedValue = snmp.PduVal2Int64(pdu)
			s.CurTime = now
			s.Scale()
			s.convertFromInteger()
			s.Valid = true
		}
```

Is the issue that you're calling the scale function *before* you're running the conversion function? In my case I want to convert the value to a float and THEN scale it, not the other way around.

**Comment by jensenja** [comment]
As a test I modified the case for the INTEGER DataSrcType in snmpmetric.go:

```
case "INTEGER", "Integer32":
		s.SetRawData = func(pdu gosnmp.SnmpPDU, now time.Time) {
			s.CookedValue = snmp.PduVal2Int64(pdu)
			s.CurTime = now
			s.convertFromInteger()
			s.Scale()
			s.Valid = true
		}
```

I flip-flopped the conversion and scale order, and now the behavior is as I would expect - with no errors and the proper float value stored in InfluxDB:

```
time="2020-03-14 18:41:00" level=info msg="SNMPDEVICE [RPP-2A1-01] Init gather cycle mode Concurrent [ true ]"
time="2020-03-14 18:41:00" level=info msg="MEASUREMENT [rpp_panel_metrics] Not Oid CONDITIONEVAL metrics exist on this measurement"
time="2020-03-14 18:41:00" level=info msg="MEASUREMENT [rpp_panel_metrics] Not EVAL metrics exist on  this measurement"
time="2020-03-14 18:41:00" level=info msg="STATS SNMP GET: snmp polling took [0.120884 seconds] SNMP: Gets [204] , Processed [204], Errors [0]"
time="2020-03-14 18:41:00" level=info msg="STATS SNMP FILTER: filter polling took [0.000000 seconds] "
time="2020-03-14 18:41:00" level=info msg="STATS INFLUX: influx send took [0.000082 seconds]"
```

InfluxDB:

```
> select panel1_total_power_factor from rpp_panel_metrics
name: rpp_panel_metrics
time                panel1_total_power_factor
----                -------------------------
1584211080000000000 95.80000000000001
1584211140000000000 97.2
1584211200000000000 97.2
1584211260000000000 97.2
```

**Comment by toni-moreno** [comment]
Hi @jensenja thank you for all this test and diagnostics.

I would like to fix it but prior we should think a few things.

1) I'm worried if the issue is the leak of a "case uint32"  in the `Scale()` instead of the compute order change. Could you test please?

2) if you're right , could this changes be also applied in other type conversions?

**Comment by jensenja** [comment]
Hi @toni-moreno I don't think that the lack of an int64 case is the real root of the issue - even if the case were added in, the current program flow dictates that the value is converted after it's scaled - which breaks use cases like mine and @ntwish when we need to add decimal places to the value by scaling it by 0.1 or 0.001 etc. The real issue isn't so much "conversion to floating point" because when an integer would be converted it's just adding ".0" to the end of it, as opposed to what we need which is changing a value from `12345` to `12.345`.

I was hesitant to submit a PR for this because I realize it's an edge case - with that said, was there a specific design decision in mind when doing the scaling before the conversion, or was the conversion done after the scaling because conversion was a newer feature? Thanks.

**Comment by toni-moreno** [comment]
Whe we did the conversion we realized that scale in the same could be a breaking change as you can see here https://github.com/toni-moreno/snmpcollector/blob/master/CHANGELOG.md#breakin.g-changes-1.

If we add a int64 option in the Scale()

 ```golang
case uint64:
		//should return Integer
              val := (s.cfg.Scale * float64(s.CookedValue.(int64))) + s.cfg.Shift
	       s.CookedValue = int64(math.Round(val))
```
with 12345 and 0.001 scaled you will have 12 as final value ( will lost the decimals).

There is a workaround.

You could define a new snmpmetric of type STRINGEVAL and use the other to get the final value as float    , in this case  extradata= `  ORIGINAL * 0.001 `, and you should exclude ORIGINAL to send to the database. In fact we would like deprecate the "Scale/Shift" option in favour of the STRINGEVAL option much more flexible.

A possible solution is to add a new "checkbox" in the snmpconfig metric option to select where you would like scaling before or after conversion... 

I  invite you to submit a PR with the ' int64 case`  currently not supported and also with the new option `scale_after_conversion`( default false and backwards compatible with current version).

**Comment by jensenja** [comment]
> with 12345 and 0.001 scaled you will have 12 as final value ( will lost the decimals).

Right - this isn't what we're trying to achieve at all.

> You could define a new snmpmetric of type STRINGEVAL and use the other to get the final value as float , in this case extradata= ORIGINAL * 0.001, and you should exclude ORIGINAL to send to the database.

I had actually considered this but opted to not implement because I have 204 individual SNMP metrics as SNMPv1 GET's for single scalar values - this would then expand out to 408 SNMP metrics, which adds quite a bit of bloat to my configuration and didn't seem ideal. I don't have a problem with taking this route though if it comes down to it.

> A possible solution is to add a new "checkbox" in the snmpconfig metric option to select where you would like scaling before or after conversion...

I had also thought about this as well but I wasn't sure if it would only serve to confuse the end user, as opposed to thinking about *why* we're not scaling after converting as opposed to the other way around.

Thanks for your responsiveness on this issue by the way!

**Comment by sbengo** [comment]
fixed in #448
